borsh-derive-internal/src/attribute_helpers.rs

```diff
@@ -18,3 +23,120 @@ pub fn contains_initialize_with(attrs: &[Attribute]) -> Option<Path> {
+                                derive_input.ident.span(),
+                                "borsh (use_discriminant=<bool>) does not support structs",
+                            )
+                            .to_compile_error(),
+                        ));
+                    }
+
+                    let value = value.to_token_stream().to_string();
+                    return match value.as_str() {
+                        "true" => Ok(Some(true)),
+                        "false" => Ok(Some(false)),
+                        _ => {
+                            return Err(TokenStream::from(
+                                syn::Error::new(
+                                    derive_input.ident.span(),
+                                    "`use_discriminant` accept only `true` or `false`",
+                                )
+                                .to_compile_error(),
+                            ));
+                        }
+                    };
+                }
+            }
+        }
+    }
+    Ok(None)
+}
+#[cfg(test)]
+mod tests {
+
+    use super::*;
+    use quote::quote;
+    #[test]
+    fn test_check_use_discriminant() {
+        let item_enum: DeriveInput = syn::parse2(quote! {
+            #[derive(BorshDeserialize, Debug)]
+            #[borsh(use_discriminant = false)]
+            enum AWithUseDiscriminantFalse {
+                X,
+                Y,
+            }
+        })
+        .unwrap();
+        let actual = check_use_discriminant(item_enum).unwrap();
+
+        assert_eq!(actual, Some(false));
+    }
+
+    #[test]
+    fn test_check_use_discriminant_true() {
+        let item_enum: DeriveInput = syn::parse2(quote! {
+            #[derive(BorshDeserialize, Debug)]
+            #[borsh(use_discriminant = true)]
+            enum AWithUseDiscriminantFalse {
+                X,
+                Y,
+            }
+        })
+        .unwrap();
+        let actual = check_use_discriminant(item_enum).unwrap();
+
+        assert_eq!(actual, Some(true));
+    }
+
+    #[test]
+    #[should_panic]
+    fn test_check_use_discriminant_wrong_value() {
+        let item_enum: DeriveInput = syn::parse2(quote! {
+            #[derive(BorshDeserialize, Debug)]
+            #[borsh(use_discriminant = 111)]
+            enum AWithUseDiscriminantFalse {
+                X,
+                Y,
+            }
+        })
+        .unwrap();
+        let actual = check_use_discriminant(item_enum);
+        assert!(actual.is_err());
+    }
+    #[test]
+    #[should_panic]
+    fn test_check_use_discriminant_on_struct() {
+        let item_enum: DeriveInput = syn::parse2(quote! {
+            #[derive(BorshDeserialize, Debug)]
+            #[borsh(use_discriminant = false)]
+            struct AWithUseDiscriminantFalse {
+                X,
+                Y,
+            }
+        })
+        .unwrap();
+        let actual = check_use_discriminant(item_enum);
+        assert!(actual.is_err());
+    }
+}
```

Comment:
Please add 2 tests with `check_item_attributes(&item_enum)`:

1. that using `#[borsh_skip]` on the whole struct/enum is error
2. that using `#[borsh(invalid)` on the whole struct/enum is error